### PR TITLE
[Product] Product variant generation

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets |  |
| License         | MIT |

Based on #5118

### CODE REVIEW

File: features/product/managing_product_variants/generating_product_variant.feature [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        Given the store is available in "English (United States)"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has option "Taste" with values "Orange" and "Melon"
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Generating a product variant for product without variants
+        Given I want to generate new variants for this product
+        When I specify that the 1st variant has 5 items in stock, is identified by "WYBOROWA_ORANGE" code and costs "$90"
+        And I specify that the 2st variant has 15 items in stock, is identified by "WYBOROWA_MELON" code and costs "$95"
```
Comment (Arminek) [inline features/product/managing_product_variants/generating_product_variant.feature]: <s>**st**</s> nd

File: features/product/managing_product_variants/generating_product_variant.feature [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        And I click generate
+        Then I should be notified that it has been successfully edited
```
Comment (Arminek) [inline features/product/managing_product_variants/generating_product_variant.feature]: <s>edited</s> generated?

File: features/product/managing_product_variants/generating_product_variant.feature [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        When I specify that the 2nd variant has 5 items in stock, is identified by "WYBOROWA_ORANGE" code and costs "$90"
+        And I click generate
+        Then I should be notified that it has been successfully edited
+        And this product should contain 2 product variants
+
+    @ui
+    Scenario: Generating the rest of product variants for product with at least one
+        Given this product is available in "Orange" taste priced at "$90.00"
+        And I want to generate new variants for this product
+        When I specify that the 2nd variant has 15 items in stock, is identified by "WYBOROWA_MELON" code and costs "$95"
```
Comment (Arminek) [inline features/product/managing_product_variants/generating_product_variant.feature]: I am not sure that stock level is relevant in this case. WDYT?

File: src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return 'sylius_admin_product_variant_generate';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getDefinedElements()
+    {
+        return array_merge(parent::getDefinedElements(), []);
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php]: Do we have to override this method if no element is defined?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_variant.yml [hunk 5]
```diff
@@ -73,3 +73,21 @@ sylius_admin_product_variant_delete:
...
+            section: admin
+            redirect:
+                route: sylius_admin_product_variant_index
+                parameters: { productId: $productId }
+            form:
+                type: sylius_product_variant_generation
+            repository:
+                method: find
+                arguments: [$productId]
+            permission: true
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_variant.yml]: I think it's the default value?

File: src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Generate/_form.html.twig [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<div class="ui one column stackable grid">
+    {%- for variant in form.variants %}
+        <div class="column">
+            <div class="ui segment">
+                    {{- form_row(variant.name) -}}
+                    {{- form_row(variant.onHand) -}}
+                    {{- form_row(variant.code) -}}
+                    {{- form_row(variant.price) -}}
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Generate/_form.html.twig]: Wrong indent.

File: src/Sylius/Bundle/ProductBundle/Resources/config/services.xml [hunk 7]
```diff
@@ -24,12 +24,15 @@
...
         <service id="sylius.validator.product_variant_combination" class="Sylius\Bundle\ProductBundle\Validator\ProductVariantCombinationValidator">
+            <argument type="service" id="sylius.checker.product_variants_parity_checker" />
             <tag name="validator.constraint_validator" alias="sylius.validator.product_variant_combination" />
         </service>
 
         <service id="sylius.generator.product_variant" class="Sylius\Component\Product\Generator\ProductVariantGenerator">
             <argument type="service" id="sylius.factory.product_variant" />
+            <argument type="service" id="sylius.checker.product_variants_parity_checker" />
         </service>
+        <service id="sylius.checker.product_variants_parity_checker" class="Sylius\Component\Product\Checker\ProductVariantsParityChecker" />
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Resources/config/services.xml]: `"sylius.checker.product_variants_parity_checker"` => `"sylius.checker.product_variants_parity"`?

File: src/Sylius/Component/Product/Checker/ProductVariantsParityChecker.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function checkParity(ProductVariantInterface $variant, ProductInterface $product)
+    {
+        foreach ($product->getVariants() as $existingVariant) {
+            // This check is require, because this function has to look for any other different variant with same option values set
+            if ($variant === $existingVariant) {
+                continue;
+            }
+
+            if (count($variant->getOptionValues()) !== count($product->getOptions())) {
```
Comment (Zales0123) [inline src/Sylius/Component/Product/Checker/ProductVariantsParityChecker.php]: Why not combine this two conditions into one?

File: src/Sylius/Component/Product/spec/Generator/ProductVariantGeneratorSpec.php [hunk 9]
```diff
@@ -26,9 +27,9 @@
  */
 final class ProductVariantGeneratorSpec extends ObjectBehavior
 {
-    function let(ProductVariantFactoryInterface $productVariantFactory)
+    function let(ProductVariantFactoryInterface $productVariantFactory, ProductVariantsParityCheckerInterface $variantsParityChecker)
```
Comment (Zales0123) [inline src/Sylius/Component/Product/spec/Generator/ProductVariantGeneratorSpec.php]: A little bit too long line.

File: src/Sylius/Component/Product/spec/Generator/ProductVariantGeneratorSpec.php [hunk 10]
```diff
@@ -75,6 +79,36 @@ function it_generates_variants_for_every_value_of_an_objects_single_option(
         $this->generate($productVariable);
     }
 
+    function it_does_not_generate_variant_if_given_variant_exist(
```
Comment (Zales0123) [inline src/Sylius/Component/Product/spec/Generator/ProductVariantGeneratorSpec.php]: ``it_does_not_generate_variant_if_given_variant_exists``

File: src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php (hunk 4 above)
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php]: I will define elements. It is just a WIP

File: src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Generate/_form.html.twig (hunk 6 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Generate/_form.html.twig]: Does not exist anymore

**Comment by pjedrzejewski** [comment]
Thanks Łukasz! :)

**Pull request merged**